sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -349,11 +349,19 @@ abstract class Optimizer(catalogManager: CatalogManager)
  */
 object EliminateDistinct extends Rule[LogicalPlan] {
   override def apply(plan: LogicalPlan): LogicalPlan = plan transformExpressions  {
-    case ae: AggregateExpression if ae.isDistinct =>
-      ae.aggregateFunction match {
-        case _: Max | _: Min => ae.copy(isDistinct = false)
-        case _ => ae
-      }
+    case ae: AggregateExpression if ae.isDistinct && isDuplicateAgnostic(ae.aggregateFunction) =>
+      ae.copy(isDistinct = false)
+  }
+
+  private def isDuplicateAgnostic(af: AggregateFunction): Boolean = af match {
+    case _: Max => true
+    case _: Min => true
+    case _: BitAndAgg => true
+    case _: BitOrAgg => true
+    case _: First => true
```

Comment:
`First` and `Last` are non-deterministic, does this matter?

Reply:
In another PR I try to argue that they should be deterministic: #29810
TLDR: An analogous aggregator would be  sum on float and double datatype - its result does depend on the order of its inputs, but is deterministic.

Reply:
If we finally make `First` and `Last` deterministic, then I guess they need to be removed from `EliminateDistinct`?

Reply:
Let's exclude First/Last here before #29810 is merged

Reply:
Done